Asset Tracker v2: Add modem module docs

jtguggedal opened the pull request:
This PR adds documentation for the modem module.

---

In addition it adds to minor patches:

applications: asset_tracker_v2: Unify data type checks 

The checks for requested data types are unified into one common
function to avoid code repetition and ease maintanance.

---

applications: asset_tracker_v2: Explicitly disable PSM if not enabled 

If the user sets CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES=n,
it's reasonable to expect that PSM would not be requested by the
device.
However, the nRF9160 modem stores PSM request information to NVM,
so PSM must explicitly be disabled before the modem is put into
normal mode to be sure that PSM is not requested by the device.

This patch adds such logic to the modem module.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6480
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

coderbyheart commented:
@jtguggedal Is it intentional that it also contains the GPS->GNSS change?

jtguggedal commented:
> @jtguggedal Is it intentional that it also contains the GPS->GNSS change?

Yes, I rebased on top of that PR as it was about to be merged, wanted to get the merge conflict resolving out of the way.
The GPS -> GNSS PR is merged now, and I've rebased on top of main, so this PR now contains only relevant changes

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+Features
+********
+
+This section documents the various features implemented by the module.
```
```suggestion
This section documents the features implemented by the module.
```

umapraseeda replied:
fixed

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+
+The module uses the :ref:`lte_lc_readme` library to control and monitor the LTE connection.
+The module registers an event handler with :ref:`lte_lc_readme` to receive updates from the modem about the state of the LTE connection and network information.
+If you use :ref:`liblwm2m_carrier_readme`, please refer to :ref:`modem_module_carrier_lib` for additional information on LTE connection handling.
```
```suggestion
If you use the :ref:`liblwm2m_carrier_readme` library, refer to :ref:`modem_module_carrier_lib` for additional information on LTE connection handling.
```

umapraseeda replied:
fixed

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+The module registers an event handler with :ref:`lte_lc_readme` to receive updates from the modem about the state of the LTE connection and network information.
+If you use :ref:`liblwm2m_carrier_readme`, please refer to :ref:`modem_module_carrier_lib` for additional information on LTE connection handling.
+
+In its default configuration, the modem module will attempt to establish an LTE connection immediately when it is initialized.
```
```suggestion
In its default configuration, the modem module makes an attempt to establish an LTE connection immediately when it is initialized.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+* ``APP_DATA_MODEM_DYNAMIC`` - Dynamic modem data, such as Cell ID, tracking area code, RSRP, IP address and PLMN (MCCMNC).
+* ``APP_DATA_BATTERY`` - Voltage of the the modem's power supply.
+
+The module uses :ref:`modem_info_readme` to  acquire information about the modem, LTE network and the modem's power supply.
```
```suggestion
The module uses :ref:`modem_info_readme` to acquire information about the modem, LTE network and the modem's power supply.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+The module uses :ref:`modem_info_readme` to  acquire information about the modem, LTE network and the modem's power supply.
+The response for the three different data types are sent as separate events:
+
+* Static data as ``MODEM_EVT_MODEM_STATIC_DATA_READY``:
+* Dynamic data as ``MODEM_EVT_MODEM_DYNAMIC_DATA_READY``:
```
```suggestion
* Static data as ``MODEM_EVT_MODEM_STATIC_DATA_READY``
* Dynamic data as ``MODEM_EVT_MODEM_DYNAMIC_DATA_READY``
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+Neighbor cell measurements
+==========================
+
+Neighbor measurements can be requested by the application by sending an ``APP_EVT_DATA_GET`` event where ``APP_DATA_NEIGHBOR_CELLS`` is part of the requested data list.
```
```suggestion
Neighbor cell measurements can be requested by the application by sending an ``APP_EVT_DATA_GET`` event where ``APP_DATA_NEIGHBOR_CELLS`` is part of the requested data list.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+==========================
+
+Neighbor measurements can be requested by the application by sending an ``APP_EVT_DATA_GET`` event where ``APP_DATA_NEIGHBOR_CELLS`` is part of the requested data list.
+Upon reception of that event, the modem module uses the LTE link controller to start a neighbor search of type ``LTE_LC_NEIGHBOR_SEARCH_TYPE_DEFAULT``. You can refer to :ref:`lte_lc_readme` for more details on the available search types.
```
```suggestion
Upon reception of that event, the modem module uses the LTE link controller to start a neighbor cell search of type ``LTE_LC_NEIGHBOR_SEARCH_TYPE_DEFAULT``. You can refer to :ref:`lte_lc_readme` for more details on the available search types.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+The modem module has an internal state machine with the following states:
+
+  * ``STATE_INIT`` - The initial state of the module in which it awaits the modem to be initialized.
+    This module will enter this state only if the :ref:`liblwm2m_carrier_readme` library is enabled, as the modem is automatically initialized otherwise.
```
```suggestion
    This module will enter this state only if the :ref:`liblwm2m_carrier_readme` library is enabled, because the modem is automatically initialized otherwise.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+
+  * ``STATE_INIT`` - The initial state of the module in which it awaits the modem to be initialized.
+    This module will enter this state only if the :ref:`liblwm2m_carrier_readme` library is enabled, as the modem is automatically initialized otherwise.
+  * ``STATE_DISCONNECTED`` - The module has performed all required initialization and is ready to establish an LTE connection.
```
```suggestion
  * ``STATE_DISCONNECTED`` - The module has performed all required initializations and is ready to establish an LTE connection.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+If the modem module receives an ``APP_EVT_LTE_DISCONNECT`` event, it will instruct the modem to detach from the network, and a subsequent ``MODEM_EVT_LTE_DISCONNECTED`` can be expected when it's done.
+
+To configure the LTE connection parameters, refer to the :ref:`lte_lc_readme` Kconfig options.
+These options can be used to select for instance which access technology to use (LTE-M or NB-IoT), default values for PSM timers and more.
```
```suggestion
These options can be used to select for instance which access technology to use (LTE-M or NB-IoT), default values for power savings mode (PSM) timers and more.
```

umapraseeda replied:
fixed

coderbyheart commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+* :kconfig:`MODEM_SEND_ALL_SAMPLED_DATA`
+
+The module will by default automatically request Power Saving Mode (PSM) from the LTE network.
+If granted by the netwok, PSM will reduce the modem's power consumption. Note that the device is not reachable from the cloud when it is in PSM. The device exits PSM whenever the application sends data or the configured PSM TAU interval has passed.
```
```suggestion
If granted by the netwok, PSM will reduce the modem's power consumption. Note that the device is not reachable from the cloud when it is in PSM. The device exits PSM whenever the application sends data or the configured PSM TAU (Tracking Area Update) interval has passed.
```

umapraseeda replied:
fixed

simensrostad commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+In its default configuration, the modem module will attempt to establish an LTE connection immediately when it is initialized.
+There is no timeout on the connection attempt, so the modem will search for a suitable LTE network until it finds one.
+
+When the device is registered to an LTE network, the LTE connection is considered established, and a ``MODEM_EVT_LTE_CONNECTED`` event is sent.
```
All events that are sent out by modules in the app can be linked to their respective header file definitions using the suggested syntax.

umapraseeda replied:
Fixed

simensrostad commented on applications/asset_tracker_v2/src/modules/modem_module.c:
```diff
@@ -725,13 +690,17 @@ static int configure_low_power(void)
 	}
 
-	LOG_DBG("PSM requested");
+	if (enable) {
```
```suggestion
	if (IS_ENABLED(CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES)) {
```

jtguggedal replied:
Fixed

umapraseeda commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,149 @@
+*****************
+
+| Header file: :file:`asset_tracker_v2/src/events/modem_module_event.h`
+| Source files: :file:`asset_tracker_v2/src/events/modem_module_event.c`
```
added

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,163 @@
+
+CONFIG_MODEM_SEND_ALL_SAMPLED_DATA - Configuration for sending all sampled data
+   By default, the modem module sends only events with sampled data that has changed since the last sampling.
+   To send unchanged data also, enable :ref:`CONFIG_MODEM_SEND_ALL_SAMPLED_DATA <CONFIG_MODEM_SEND_ALL_SAMPLED_DATA>`.
```
Not sure whether it is reasonable to link to the Kconfig option itself.

umapraseeda replied:
removed the link.

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,163 @@
+   If PSM is granted by the network, it results in reduction of the modem's power consumption.
+   Note that the device is not reachable from the cloud when it is in PSM.
+   The device exits PSM whenever the application sends data, or the configured PSM TAU (Tracking Area Update) interval has passed.
+   To not request PSM from the network, disable :ref:`CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES <CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES>`.
```
Also this links to the Kconfig option itself.

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,163 @@
+
+CONFIG_MODEM_SEND_ALL_SAMPLED_DATA - Configuration for sending all sampled data
+   By default, the modem module sends only events with sampled data that has changed since the last sampling.
+   To send unchanged data also, enable CONFIG_MODEM_SEND_ALL_SAMPLED_DATA.
```
Can we just say: "..., enable this option"?

peknis commented on applications/asset_tracker_v2/doc/modem_module.rst:
```diff
@@ -0,0 +1,163 @@
+   If PSM is granted by the network, it results in reduction of the modem's power consumption.
+   Note that the device is not reachable from the cloud when it is in PSM.
+   The device exits PSM whenever the application sends data, or the configured PSM TAU (Tracking Area Update) interval has passed.
+   To not request PSM from the network, disable CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES.
```
Same here, "..., disable this option".

jtguggedal commented:
@umapraseeda  Thanks for the review, I have squashed in the changes

jtguggedal commented on applications/asset_tracker_v2/src/modules/modem_module.c:
```diff
@@ -926,8 +896,9 @@ static void on_all_states(struct modem_msg_data *msg)
-						  msg->module.app.count)) {
+		if (data_type_is_requested(msg->module.app.data_list,
+					   msg->module.app.count,
+					   APP_DATA_NEIGHBOR_CELLS) {
```
Fixed
